pkg/audit/stats_reporter.go

```diff
@@ -2,111 +2,142 @@ package audit
+	)
+	if err != nil {
+		panic(err)
 	}
 
-	return r.report(ctx, violationsM.M(v))
+	view.Register(sdkmetric.NewView(
+		sdkmetric.Instrument{Name: auditDurationMetricName},
+		sdkmetric.Stream{
+			Aggregation: sdkmetric.AggregationExplicitBucketHistogram{
+				Boundaries: []float64{1 * 60, 3 * 60, 5 * 60, 10 * 60, 15 * 60, 20 * 60, 40 * 60, 80 * 60, 160 * 60, 320 * 60},
+			},
+		},
+	))
 }
 
-func (r *reporter) reportLatency(d time.Duration) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
+func (r *reporter) registerCallback() error {
+	_, err1 := meter.RegisterCallback(r.observeTotalViolations, violationsM)
+	_, err2 := meter.RegisterCallback(r.observeRunEnd, lastRunEndTimeM)
+	_, err3 := meter.RegisterCallback(r.observeRunStart, lastRunStartTimeM)
+	return errors.Join(err1, err2, err3)
+}
+
+func (r *reporter) observeTotalViolations(_ context.Context, o metric.Observer) error {
+	for k, v := range r.totalViolationsPerEnforcementAction {
+		o.ObserveInt64(violationsM, v, metric.WithAttributes(attribute.String(enforcementActionKey, string(k))))
 	}
+	return nil
+}
 
-	return r.report(ctx, auditDurationM.M(d.Seconds()))
+func (r *reporter) reportTotalViolations(enforcementAction util.EnforcementAction, v int64) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.totalViolationsPerEnforcementAction == nil {
+		r.totalViolationsPerEnforcementAction = make(map[util.EnforcementAction]int64)
+	}
+	r.totalViolationsPerEnforcementAction[enforcementAction] = v
+	return nil
 }
 
 func (r *reporter) reportRunStart(t time.Time) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
-	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.startTime = t
+	return nil
+}
 
-	val := float64(t.Unix())
-	return metrics.Record(ctx, lastRunStartTimeM.M(val))
+func (r *reporter) reportLatency(d time.Duration) error {
+	auditDurationM.Record(context.Background(), d.Seconds())
+	return nil
 }
 
 func (r *reporter) reportRunEnd(t time.Time) error {
-	ctx, err := tag.New(context.Background())
-	if err != nil {
-		return err
-	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.endTime = t
+	return nil
+}
+
+func (r *reporter) observeRunStart(_ context.Context, o metric.Observer) error {
+	o.ObserveFloat64(lastRunStartTimeM, float64(r.startTime.Unix()))
+	return nil
+}
 
-	val := float64(t.Unix())
-	return metrics.Record(ctx, lastRunEndTimeM.M(val))
+func (r *reporter) observeRunEnd(_ context.Context, o metric.Observer) error {
+	o.ObserveFloat64(lastRunEndTimeM, float64(r.endTime.Unix()))
+	return nil
 }
 
 // newStatsReporter creates a reporter for audit metrics.
 func newStatsReporter() (*reporter, error) {
-	return &reporter{}, nil
+	r := &reporter{}
+	return r, r.registerCallback()
 }
 
-type reporter struct{}
-
-func (r *reporter) report(ctx context.Context, m stats.Measurement) error {
-	return metrics.Record(ctx, m)
+type reporter struct {
+	mu                                  sync.RWMutex
```

Comment:
If we wind up with a lot of lock contention, it may be worth splitting this out into one-mutex-per-metric (or using atomic variables where possible), but probably worth seeing if that's necessary first.